### PR TITLE
Archive page with all issues - #247

### PR DESCRIPTION
**Pull request by Niubi90**
First PR in opensource :)

### CODE REVIEW

File: src/main/java/com/jvm_bloggers/frontend/newsletter_issue/NewsletterIssueDtoService.java [hunk 1]
```diff
@@ -20,19 +20,24 @@
...
     }
 
     public List<NewsletterIssueDto> findTop5ByOrderByPublishedDateDesc() {
         return newsletterIssueRepository.findTop5ByOrderByPublishedDateDesc()
             .stream().map(newsletterIssueDtoBuilder::build).collect(Collectors.toList());
     }
+
+    public List<NewsletterIssueDto> findAllByOrderByPublishedDateDesc() {
+        return newsletterIssueRepository.findAllByOrderByPublishedDateDesc().stream()
+            .map(newsletterIssueDtoBuilder::build).collect(Collectors.toList());
```
Comment (mszarlinski) [inline src/main/java/com/jvm_bloggers/frontend/newsletter_issue/NewsletterIssueDtoService.java]: Maybe start each transformation on a stream from a new line to enhance readability. I would also use a static import for `Collectors#toList`.

File: src/main/java/com/jvm_bloggers/frontend/archiwum_issue/ArchiwumIssuePage.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        .ofPattern("MM/yyyy");
+
+    @SpringBean
+    private NewsletterIssueDtoService newsletterIssueDtoService;
+
+    public ArchiwumIssuePage() {
+        HashMap<String, List<Link>> archiwumIssuesGroup = 
+            (HashMap<String, List<Link>>) newsletterIssueDtoService
+            .findAllByOrderByPublishedDateDesc().stream()
+            .collect(Collectors.groupingBy(this::getArchiwumIssueGroup,
```
Comment (mszarlinski) [inline src/main/java/com/jvm_bloggers/frontend/archiwum_issue/ArchiwumIssuePage.java]: I'm too retarded to understand what is happening here. Please extract a grouping collector to method and give it some descriptive name.

File: src/main/java/com/jvm_bloggers/frontend/archiwum_issue/ArchiwumIssuePage.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        .ofPattern("dd/MM/yyyy");
+    static final DateTimeFormatter ARCHIWUM_ISSSUE_FORMATTER = DateTimeFormatter
+        .ofPattern("MM/yyyy");
+
+    @SpringBean
+    private NewsletterIssueDtoService newsletterIssueDtoService;
+
+    public ArchiwumIssuePage() {
+        HashMap<String, List<Link>> archiwumIssuesGroup = 
+            (HashMap<String, List<Link>>) newsletterIssueDtoService
```
Comment (mszarlinski) [inline src/main/java/com/jvm_bloggers/frontend/archiwum_issue/ArchiwumIssuePage.java]: Why casting to particular `java.util.Map` implementation?

Comment (Niubi90) [inline src/main/java/com/jvm_bloggers/frontend/archiwum_issue/ArchiwumIssuePage.java]: Yeah, right it isn't necessary.

File: src/main/java/com/jvm_bloggers/frontend/newsletter_issue/NewsletterIssueDtoService.java (hunk 1 above)
Comment (Niubi90) [inline src/main/java/com/jvm_bloggers/frontend/newsletter_issue/NewsletterIssueDtoService.java]: @mszarlinski Ok, i will put each transformation on a stream from a new line. Why should i use static import to Collectors#toList?

File: src/main/java/com/jvm_bloggers/frontend/archiwum_issue/ArchiwumIssuePage.java (hunk 2 above)
Comment (Niubi90) [inline src/main/java/com/jvm_bloggers/frontend/archiwum_issue/ArchiwumIssuePage.java]: I did some method extracting. You can see this in next commit.

File: src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    private Map<String, List<Link>> createArchivedMonthGroups(
+        NewsletterIssueDtoService newsletterIssueDtoService) {
+        return newsletterIssueDtoService
+            .findAllByOrderByPublishedDateDesc().stream()
+            .collect(Collectors.groupingBy(this::getArchivedIssueGroup,
+                Collectors.mapping(this::getLink, Collectors.toList())));
+    }
+
+    private Map<String, List<Link>> sortMapByKey(Map<String, List<Link>> map) {
+        Map<String, List<Link>> sortedMap = new LinkedHashMap<String, List<Link>>();
```
Comment (airborn) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java]: Diamond operator could be used here.

File: src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        map.entrySet().stream()
+            .sorted(Map.Entry.<String, List<Link>>comparingByKey()
+                .reversed())
+            .forEachOrdered(e -> sortedMap.put(e.getKey(), e.getValue()));
```
Comment (airborn) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java]: Do we really need to copy the map we've just created? Couldn't we simply use overloaded `Collectors.groupingBy` that accepts `Suplier<TreeMap> `with appropriate `Comparator`?

File: src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        return sortedMap;
+    }
+
+    private String getArchivedIssueGroup(NewsletterIssueDto issue) {
+        return ARCHIVED_ISSSUE_FORMATTER.format(issue.publishedDate);
+    }
+
+    private Link getLink(NewsletterIssueDto issue) {
+        return (Link) new BookmarkablePageLink<>("issueLink", NewsletterIssuePage.class,
```
Comment (airborn) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java]: Is this casting really needed?

Comment (Niubi90) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java]: @airborn
Hmm, compiler tells me:
Type mismatch: cannot convert from AbstractLink to Link, when haven't casting.
So, I guess it should be.

File: src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java (hunk 4 above)
Comment (Niubi90) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java]: Method was deleted.

File: src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    private static final DateTimeFormatter PUBLISHED_DATE_FORMATTER = DateTimeFormatter
+        .ofPattern("dd/MM/yyyy");
+    private static final DateTimeFormatter ARCHIVED_ISSSUE_FORMATTER = DateTimeFormatter
+        .ofPattern("MM/yyyy");
+
+    @SpringBean
+    private NewsletterIssueDtoService newsletterIssueDtoService;
+
+    public ArchivedIssuePage() {
+        SortedMap<String, List<Link>> archiwumIssuesGroup = createArchivedMonthGroups(
```
Comment (Niubi90) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java]: Will change

File: src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        add(new ArchivedIssuePanel(ARCHIVED_ISSUE_PANEL_ID, archiwumIssuesGroup));
+    }
+
+    private SortedMap<String, List<Link>> createArchivedMonthGroups(
+        NewsletterIssueDtoService newsletterIssueDtoService) {
+        return newsletterIssueDtoService
+            .findAllByOrderByPublishedDateDesc().stream()
+            .collect(Collectors.groupingBy(
+                this::getArchivedIssueGroup,
+                () -> new TreeMap<String, List<Link>>(Comparator.reverseOrder()),
```
Comment (Niubi90) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java]: done :)

File: src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+                () -> new TreeMap<>(Comparator.reverseOrder()),
+                Collectors.mapping(this::getLink, Collectors.toList())));
+    }
+
+    private String getArchivedIssueGroup(NewsletterIssueDto issue) {
+        return ARCHIVED_ISSUE_FORMATTER.format(issue.publishedDate);
+    }
+
+    private Link getLink(NewsletterIssueDto issue) {
+        return (Link) new BookmarkablePageLink<>("issueLink", NewsletterIssuePage.class,
```
Comment (goostleek) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java]: Compilation warning: 
> Link is a raw type. References to generic type Link<T> should be parameterized

File: src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+                this::getArchivedIssueGroup,
+                () -> new TreeMap<>(Comparator.reverseOrder()),
+                Collectors.mapping(this::getLink, Collectors.toList())));
+    }
+
+    private String getArchivedIssueGroup(NewsletterIssueDto issue) {
+        return ARCHIVED_ISSUE_FORMATTER.format(issue.publishedDate);
+    }
+
+    private Link getLink(NewsletterIssueDto issue) {
```
Comment (goostleek) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java]: Compilation warning: 
> Link is a raw type. References to generic type Link<T> should be parameterized

File: src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    private static final DateTimeFormatter PUBLISHED_DATE_FORMATTER = DateTimeFormatter
+        .ofPattern("dd/MM/yyyy");
+    private static final DateTimeFormatter ARCHIVED_ISSUE_FORMATTER = DateTimeFormatter
+        .ofPattern("MM/yyyy");
+
+    @SpringBean
+    private NewsletterIssueDtoService newsletterIssueDtoService;
+
+    public ArchivedIssuePage() {
+        SortedMap<String, List<Link>> archivedIssuesGroup = createArchivedMonthGroups(
```
Comment (goostleek) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java]: Compilation warning: 
> Link is a raw type. References to generic type Link<T> should be parameterized

File: src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+@MountPath("archivedIssues")
+public class ArchivedIssuePage extends AbstractFrontendPage {
+
+    private static final String ARCHIVED_ISSUE_PANEL_ID = "archivedIssuePanel";
+    private static final DateTimeFormatter PUBLISHED_DATE_FORMATTER = DateTimeFormatter
+        .ofPattern("dd/MM/yyyy");
+    private static final DateTimeFormatter ARCHIVED_ISSUE_FORMATTER = DateTimeFormatter
```
Comment (goostleek) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java]: Duplicated formatter. Use [`DateTimeUtilities#DATE_FORMATTER`](https://github.com/jvm-bloggers/jvm-bloggers/blob/master/src/main/java/com/jvm_bloggers/utils/DateTimeUtilities.java#L15) instead.

File: src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+        add(new ArchivedIssuePanel(ARCHIVED_ISSUE_PANEL_ID, archivedIssuesGroup));
+    }
+
+    private SortedMap<String, List<Link>> createArchivedMonthGroups(
+        NewsletterIssueDtoService newsletterIssueDtoService) {
+        return newsletterIssueDtoService
+            .findAllByOrderByPublishedDateDesc().stream()
+            .collect(Collectors.groupingBy(
+                this::getArchivedIssueGroup,
+                () -> new TreeMap<>(Comparator.reverseOrder()),
```
Comment (goostleek) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java]: The statement will not compile under Eclipse:
```
() -> new TreeMap<>(Comparator.reverseOrder())
```
the fix is:
```
() -> new TreeMap<String, List<Link>>(Comparator.reverseOrder())
```

File: src/main/java/com/jvm_bloggers/frontend/archived_issue/archived_panel/ArchivedIssuePanel.java [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+import org.apache.wicket.markup.html.link.Link;
+import org.apache.wicket.markup.html.panel.Panel;
+import org.apache.wicket.markup.repeater.RepeatingView;
+
+import java.util.List;
+import java.util.Map;
+
+public class ArchivedIssuePanel extends Panel {
+
+    public ArchivedIssuePanel(String id, Map<String, List<Link>> archivedIssueGroups) {
```
Comment (goostleek) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/archived_panel/ArchivedIssuePanel.java]: Compilation warning:

> Link is a raw type. References to generic type Link<T> should be parameterized

File: src/main/java/com/jvm_bloggers/frontend/archived_issue/archived_panel/IssuesInMonthPanel.java [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+import org.apache.wicket.markup.html.link.Link;
+import org.apache.wicket.markup.html.list.ListItem;
+import org.apache.wicket.markup.html.list.ListView;
+import org.apache.wicket.markup.html.panel.Panel;
+
+import java.util.List;
+
+public class IssuesInMonthPanel extends Panel {
+
+    public IssuesInMonthPanel(String id, String groupLabel, List<Link> monthIssues) {
```
Comment (goostleek) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/archived_panel/IssuesInMonthPanel.java]: Compilation warning:

> Link is a raw type. References to generic type Link<T> should be parameterized

File: src/main/java/com/jvm_bloggers/frontend/archived_issue/archived_panel/IssuesInMonthPanel.java [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+        super(id);
+
+        add(new Label("groupLabel", groupLabel));
+        add(new ListView<Link>("issuesList", monthIssues) {
+
+            @Override
+            protected void populateItem(ListItem<Link> item) {
```
Comment (goostleek) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/archived_panel/IssuesInMonthPanel.java]: Compilation warning:

> Link is a raw type. References to generic type Link<T> should be parameterized

File: src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java (hunk 13 above)
Comment (airborn) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java]: This is not a fix, this is walk around. Should we really care that eclipse cannot understand valid java syntax?

Comment (goostleek) [inline src/main/java/com/jvm_bloggers/frontend/archived_issue/ArchivedIssuePage.java]: I think the code should compile in Eclipse also. I can give you a reason - I'm using Eclipse :) And I think I'm not the only one in the world still using it :)

**Comment by Niubi90** [comment]
In time will work on some tests for implemented Wicket components

**Pull request merged**